Fancy wifi Table and Camera Load widget (Challenge)

hanupratap opened the pull request:
WebUI Challenge Task 1:

I have added sorting functionality based on the keys of the objects available in WiFi tab. Also, I have recreated the table to match the material theme.

NOTE : Click the header of the column to sort. Initially sorted by IN-USE key

hanupratap committed:
Added sort by key feature in WiFi tab and Improved UI in WiFi tab

hanupratap commented:
> WebUI Challenge Task 1:
> 
> I have added sorting functionality based on the keys of the objects available in WiFi tab. Also, I have recreated the table to match the material theme.
> 
> NOTE : Click the header of the column to sort. Initially sorted by IN-USE key

@rroohhh

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -1,5 +1,23 @@
+import TableHead from '@material-ui/core/TableHead';
+import TableRow from '@material-ui/core/TableRow';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
```
unused import

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -1,5 +1,23 @@
+import Paper from '@material-ui/core/Paper';
+import FormControlLabel from '@material-ui/core/FormControlLabel';
+import Switch from '@material-ui/core/Switch';
+import color from '@material-ui/core/colors/amber';
```
unused import

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+  },
+}));
+
+const EnhancedTableToolbar = (props) => {
```
`props` is not used in this function, so this argument is unnecessary

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+  );
+};
+
+EnhancedTableToolbar.propTypes = {
```
unnecessary

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+
+function descendingComparator(a, b, orderBy) {
+  let is_num = false;
+  headCells.map((element)=>{
+    if(element.id == orderBy)
+      is_num = is_num || element.numeric
+  });
```
Why not make `headCells` a dictionary with `id` as the key?

This way you can avoid this search.

hanupratap replied:
Noted

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+
+function EnhancedTable(props) {
+  const classes = useStyles();
+  const [order, setOrder] = React.useState('asc');
```
To keep consistent with the rest of the codebase please import `useState` directly.

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+                  );
+                })}
+
+              {/* {emptyRows > 0 && (
```
Please remove unnecessary commented out code.

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+          </Table>
+        </TableContainer>
+      </Paper>
+      <FormControlLabel
```
I don't see the value in adding this toggle here you can remove it for now, maybe we can globally add such a toggle to a settings in the future.

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -10,44 +28,268 @@ export class Component extends React.Component {
+  { id: 'SSID', numeric: false, disablePadding: false, label: 'SSID' },
+  { id: 'MODE', numeric: false, disablePadding: false, label: 'MODE' },
+  { id: 'CHAN', numeric: true, disablePadding: false, label: 'CHAN' },
+  { id: 'RATE', numeric: true, disablePadding: false, label: 'RATE (Mbits/sec)' },
```
If the values in the table column already have `Mbits/sec` as suffix, adding it to the table header is redundant.

Please either strip it from the values in the column, or omit it from the table header.

hanupratap replied:
Changed in new commit

rroohhh reviewed (changes requested):
Nice work so far. Please see the inline comment.

Some general notes:
1. please make sure your code is formatted properly, you can run the autoformatter using `npm run format`
2. please make sure the code is clean, no commented out code, no unused imports, etc

Finally, I think `IN-USE` could be visualized better by making the text in the row bold or changing the background of the row.

hanupratap committed:
Highlighted row and added active dot for connected network. Removed unncecesary imports and comments.

hanupratap commented:
Hello @rroohhh , thanks for suggesting changes. I have added a commit which caters to all your suggestions. Let me know if any more changes are required :)

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -1,5 +1,22 @@
+import { green, orange } from '@material-ui/core/colors';
+import Skeleton from '@material-ui/lab/Skeleton';
+import { Box } from '@material-ui/core';
+import { isNullishCoalesce } from 'typescript';
```
unused import

hanupratap replied:
Removed

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -12,42 +29,235 @@ export class Component extends React.Component {
+// Meta-Data
+
+const in_use = 'IN-USE';
+const first_column = 'IN-USE';
```
If you are creating a constant for `'IN-USE'`, please use it everywhere `'IN-USE'` is used, otherwise they could become inconsistent.

hanupratap replied:
Removed first_column, now using in_use everywhere.

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -12,42 +29,235 @@ export class Component extends React.Component {
+const first_column = 'IN-USE';
+const active_color = orange[50];
+const headCells = {
+  'IN-USE': { numeric: false, disablePadding: true, label: 'Connected' },
```
See above.

hanupratap replied:
Hey, I had a doubt in this. How can I replace IN-USE here, there is a hyphen. How can i make this consistent? I have tried making other keys with quotes but `npm run format` removes the quotes except for `IN-USE`

hanupratap replied:
Should I remove the constant and use 'IN-USE' everywhere ?

rroohhh replied:
I think having the constant is fine, but if you have a constant please use it everywhere,  (You can use it in the dictionary like `const headCells = { [in_use] : .... }`

hanupratap replied:
Done

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -12,42 +29,235 @@ export class Component extends React.Component {
+function EnhancedTable(props) {
+  const classes = useStyles();
+  const [order, setOrder] = useState('asc');
+  const [orderBy, setOrderBy] = useState(first_column);
```
Why use a constant for `first_column` and not for `'asc'`. Please be consistent in the code, I think here using a constant just for `first_column` is unnecessary.

hanupratap replied:
removed first_column, Created constant for initial direction('asc') as well.

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -12,42 +29,235 @@ export class Component extends React.Component {
-  const { data } = props;
-  if (data.length === 0) return <table />;
+
+// Meta-Data
```
Unnecessary comment (also the one below).

hanupratap replied:
Removed

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -12,42 +29,235 @@ export class Component extends React.Component {
+
+function EnhancedTableHead(props) {
+  const { classes, order, orderBy, onRequestSort } = props;
+  const createSortHandler = property => event => {
```
inlining this logic would be cleaner IMO

hanupratap replied:
Made this inline

hanupratap replied:
> Great so far, however
> 
> 1. The formatting is still not completely clean (`npm run format` still reformats some things)
> 2. What do you think about making the security column nicer using chips?
> 3. Finally `Enhanced*` are pretty generic names, I think more meaningful names would be great.
> 
> Also please see the inline comments.

Hi @rroohhh , Please check my new commit. I have made changes as per your suggestions. Also please check in the above suggestions, I have one query for you. 

Regarding recent commit, I have added chips for the security column and it looks good :). I have also checked the dependencies and updated it. Also, I have changed the colour from white to its hex equivalent(#ffff) in index.jsx, since it was causing some problems in rendering.

Do let me know what do you think, meanwhile I will start working on the load graph :)

![Selection_003](https://user-images.githubusercontent.com/51860315/111785027-17e80480-88e2-11eb-9f2d-1545ab224454.png)

rroohhh commented on package.json:
```diff
@@ -14,23 +14,24 @@
     "@material-ui/icons": "^4.5.1",
     "@reduxjs/toolkit": "^1.1.0",
-    "@types/react": "^16.9.17",
+    "@types/react": "^16.8.6 || ^17.0.0",
```
Are these actually proper version constraints? If you need `^17.0.0` that is fine, but accepting `^16.8.6` and `^17.0.0` makes no sense then?

hanupratap replied:
Changed

rroohhh reviewed (changes requested):
Great so far, however
1. The formatting is still not completely clean (`npm run format` still reformats some things)
2. What do you think about making the security column nicer using chips?
3. Finally `Enhanced*` are pretty generic names, I think more meaningful names would be great.

Also please see the inline comments.

hanupratap committed:
Added Chips for WPA, Updated dependecies, cleaned code

hanupratap commented:
Hi @rroohhh , I have made changes as per your suggestions. Also please check in the above suggestions, I have one query for you. [link to comment](https://github.com/apertus-open-source-cinema/webui/pull/11#discussion_r597658940)

Regarding recent commit, I have added chips for the security column and it looks good :). I have also checked the dependencies and updated it. Also, I have changed the colour from white to its hex equivalent(#ffff) in index.jsx, since it was causing some problems in rendering.

Do let me know what do you think, meanwhile I will start working on the load graph :)

![Selection_003](https://user-images.githubusercontent.com/51860315/111807738-491fff00-88f9-11eb-945f-4780e6947899.png)

rroohhh commented on frontend/routes/Wifi.jsx:
```diff
@@ -73,14 +70,9 @@ const headCells = {
-  const createSortHandler = property => event => {
-    onRequestSort(event, property);
-  };
+  const createSortHandler = property => event => onRequestSort(event, property);
```
By inline I meant instead of creating a function here with just a single use, just write this directly in the `onClick` handler, sorry if that was unclear!

hanupratap replied:
Done

rroohhh commented:
Looks good so far, maybe left-aligning the Security column would be cleaner. 
Otherwise, looking forward to your load graph solution :)

hanupratap committed:
Added camer load graph

hanupratap commented:
Hi @rroohhh , I have added a commit for the load graph where I have used react-charts library. Let me know what you think :)

![Screenshot from 2021-03-22 01-46-56](https://user-images.githubusercontent.com/51860315/111920599-da3bd500-8ab5-11eb-97b3-f8ca63b414f8.png)

hanupratap committed:
Minor UI Changes and code improvement

hanupratap commented:
Also, I have made changes according to your previous suggestions in my latest commit.
Waiting for your feedback :)

hanupratap committed:
Minor function name changes

rroohhh commented on frontend/components/CommandWidgets.jsx:
```diff
@@ -9,12 +9,54 @@ export function Command(props) {
-        .then(result => setOutput(result[0]))
+        .then(result => {
+          setOutput(result[0]);
+          if (updateData !== undefined) updateData(extractLoadFromString(result[0]));
```
This functionality does not belong into the `PlanCommand` widget.

rroohhh commented on frontend/components/CommandWidgets.jsx:
```diff
@@ -9,12 +9,54 @@ export function Command(props) {
+      label: 'One Minute Load',
+      data: [
+        {
+          primary: new Date(),
```
These three dates should probably all be the same, and probably the one output by the `uptime` command.

rroohhh commented on frontend/components/GraphUtils/useDevConfig.js:
```diff
@@ -0,0 +1,155 @@
+import React, { useEffect, useState } from 'react';
```
Why is this file called `useDevConfig`, if you are exporting a `useChartConfig`? Furthermore both names are not really descriptive.

rroohhh commented on frontend/components/GraphUtils/useDevConfig.js:
```diff
@@ -0,0 +1,155 @@
+  const updateData = new_data => {
+    if (new_data === undefined) return null;
+
+    const temp_data = state;
```
What do you intent to achieve with this assignment?

rroohhh commented on frontend/components/GraphUtils/useDevConfig.js:
```diff
@@ -0,0 +1,155 @@
+    const temp_data = state;
+    temp_data.data = temp_data.data.map((ele, i) => {
+      return {
+        label: new_data[i].label,
```
Is it really necessary to update the label every time?

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -0,0 +1,64 @@
+import { Box, Typography } from '@material-ui/core';
+
+export default function LoadGraph() {
+  const datasets = {
```
unused

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -0,0 +1,64 @@
+import React from 'react';
+import { Chart } from 'react-charts';
+
+import useDemoConfig from './GraphUtils/useDevConfig';
```
And here its called `useDemoConfig`, the name should atleast be consistent.

rroohhh commented on frontend/components/GraphUtils/useDevConfig.js:
```diff
@@ -0,0 +1,155 @@
+    }));
+  };
+
+  const Options = optionKeys
```
completely unused.

rroohhh reviewed (changes requested):
Good start on the load graph widget.
Some notes:
1. Its missing a legend
2. Its missing axis labels
3. It gets quite laggy when running for some time
4. The theme doesn't really match up with the rest of the site
5. Zooming / scrolling would be great

Also see the inline notes.

hanupratap committed:
Changed graph library to ApexCharts

hanupratap committed:
graph update

hanupratap committed:
Added bars icon

hanupratap committed:
formatted code

hanupratap commented:
hello @rroohhh , I have changed the library itself as it didn't offer much features. Now, I am using apex-charts and it looks better. It doesn't look jittery now. I have also added a color highlight in the graph if the load increases. Also, now we can download the data directly from the graph as csv or download the graph itself as svg.

Re-render problem: 
When we need to zoom/ pan or interact in any way with the graph it resets because it's going to re-render every second.

Suggested solution:
We can have a button which stops the live data and then we can interact with the graph, since it will not re-render. Should I work on this ?

Suggestion to improve performance (display limited points) : 
We can specify at a time how much data we need to display, for instance we can show 100 points at a time, whenever our data exceeds 100 points, we will shift our data-array. But I am not sure if this is required ? 

![camera_load](https://user-images.githubusercontent.com/51860315/112217058-3cc4da80-8c48-11eb-91fa-146c1b6dcdae.png)

hanupratap commented:
Also, I have replaced the bars string with an appropriate icon. I found those in the materialUI library. PFA - screenshot

![bars_icons](https://user-images.githubusercontent.com/51860315/112217900-12275180-8c49-11eb-9317-8b3cd83c1590.jpg)

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -1,64 +1,124 @@
       </Box>
-      <Command command="uptime" interval={2000} updateData={updateData} />
+
+      <pre style={{ backgroundColor: '#eee', overflowX: 'auto' }}>
```
Why not factor this out into something both `Command` and this widget can use?

hanupratap replied:
Hey, I have added a commit which shares data between graph and command.

rroohhh reviewed (changes requested):
Nice work,

> hello @rroohhh , I have changed the library itself as it didn't offer much features. Now, I am using apex-charts and it  looks better. It doesn't look jittery now. I have also added a color highlight in the graph if the load increases. Also, now we can download the data directly from the graph as csv or download the graph itself as svg.

Nice, I like the color highlights aswell as the csv / svg download!

> Re-render problem:
> When we need to zoom/ pan or interact in any way with the graph it resets because it's going to re-render every second.
> 
> Suggested solution:
> We can have a button which stops the live data and then we can interact with the graph, since it will not re-render. Should I work on this ?

This could work as a last resort solution, but finding something that does allow updating the data while keeping things like zoom and position would be a lot better.

> Suggestion to improve performance (display limited points) :
> We can specify at a time how much data we need to display, for instance we can show 100 points at a time, whenever our data exceeds 100 points, we will shift our data-array. But I am not sure if this is required ?

Well only displaying 100 points at a time would reduce the lag, but finding a way to atleast display like 10000 points interactively should be possible imo. For a really big number of points one could investigate some data reduction methods like averaging nearby points, but I don't think that should be necessary for only 10k points.


Futhermore, this graph is now missing a proper x axis, also it seems like it displays one tick / entry on the x axis per entry, which makes it quickly unreadable.

hanupratap committed:
Refactored command and graph widget

hanupratap committed:
Fixed graph reset on re-render

hanupratap committed:
Formatted Code

hanupratap commented:
> Nice work,
> 
> > hello @rroohhh , I have changed the library itself as it didn't offer much features. Now, I am using apex-charts and it  looks better. It doesn't look jittery now. I have also added a color highlight in the graph if the load increases. Also, now we can download the data directly from the graph as csv or download the graph itself as svg.
> 
> Nice, I like the color highlights aswell as the csv / svg download!
> 
Glad you liked it
> > Re-render problem:
> > When we need to zoom/ pan or interact in any way with the graph it resets because it's going to re-render every second.
> > Suggested solution:
> > We can have a button which stops the live data and then we can interact with the graph, since it will not re-render. Should I work on this ?
> 
> This could work as a last resort solution, but finding something that does allow updating the data while keeping things like zoom and position would be a lot better.
> 
Hey, I have fixed this issue of resetting graph on zoom/pan, please check recent commit. I am manually accessing the chart through DOM and updating the series.
> > Suggestion to improve performance (display limited points) :
> > We can specify at a time how much data we need to display, for instance we can show 100 points at a time, whenever our data exceeds 100 points, we will shift our data-array. But I am not sure if this is required ?
> 
> Well only displaying 100 points at a time would reduce the lag, but finding a way to atleast display like 10000 points interactively should be possible imo. For a really big number of points one could investigate some data reduction methods like averaging nearby points, but I don't think that should be necessary for only 10k points.
> 
Processing large number of points should not be laggy anymore, from my latest commit I am only appending the series and the graph is not redrawing the previous points.
> Futhermore, this graph is now missing a proper x axis, also it seems like it displays one tick / entry on the x axis per entry, which makes it quickly unreadable.

I am not really sure what you mean, can you please elaborate on this. I have removed the x-axis labels since they were not readable for large number of points. The labels will only be displayed when hovered.

rroohhh commented on frontend/components/RunCommand.jsx:
```diff
@@ -0,0 +1,77 @@
+    let fifteen_min = parseFloat(list[4]);
+
+    return {
+      x_value: new Date(),
```
Why generate the `Date` here instead of using the one in the output of `uptime`?

rroohhh commented on frontend/components/RunCommand.jsx:
```diff
@@ -0,0 +1,77 @@
+const graph_title = 'Load average';
+
+export default function RunCommand(props) {
+  const [state, setstate] = useState({
```
We use camelCase elsewhere, please try to follow the existing code style (`setState`)

rroohhh commented on frontend/components/RunCommand.jsx:
```diff
@@ -0,0 +1,77 @@
+    const interval_handle = setInterval(callback, props.interval);
+    callback();
+    return () => clearInterval(interval_handle);
+  }, [props.command]);
```
maybe this should also rebuild if `props.showGraph` changes

rroohhh commented on frontend/components/RunCommand.jsx:
```diff
@@ -0,0 +1,71 @@
+import { Box } from '@material-ui/core';
```
Sorry if I was unclear, I didn't mean, merge the load chart and the command output into one widget, I simply meant refactor the code is such a way that you can use the command output formatting from the normal commands for displaying the output of the `uptime` command used for the load graph.

rroohhh commented on frontend/components/LoadChartWidget.jsx:
```diff
@@ -0,0 +1,84 @@
+      data: [],
+    },
+    {
+      name: '15 mintue',
```
typo

rroohhh reviewed (changes requested):
> Hey, I have fixed this issue of resetting graph on zoom/pan, please check recent commit. I am manually accessing the chart through DOM and updating the series.

Nice, it works better now, bit it still seems to have some problems:
- panning is acting really strange and is noninterative
- if the data gets updated while selecting a interval to zoom to, the start of the interval gets reset to the left side

> Processing large number of points should not be laggy anymore, from my latest commit I am only appending the series and the graph is not redrawing the previous points.

To me it seems still pretty laggy, for example try resizing the browser windows with ~100 points for example, that takes multiple seconds on my laptop

> > Futhermore, this graph is now missing a proper x axis, also it seems like it displays one tick / entry on the x axis per entry, which makes it quickly unreadable.
>
> I am not really sure what you mean, can you please elaborate on this. I have removed the x-axis labels since they were not readable for large number of points. The labels will only be displayed when hovered.

Sorry for being unclear. What I meant was, that a increasing number does not seem like the correct x value for a timeseries, the time when the measurement was taken makes a lot more sense imo.  So it would be great to have the time on the x-axis (and as x ticks).  
Also while you have removed the x-axis labels, the ticks are still there.

hanupratap committed:
Removed ticks and added loading text for no data

hanupratap committed:
Changed case for setstate

hanupratap committed:
added showGraph in rebuild list

hanupratap committed:
Added CanvasJs Charts

hanupratap commented:
Hey @rroohhh, I have added canvasJs charts, let me know if it is performant enough. Also, I haven't completely removed the previous library yet, I'll remove unused libs once we are fixed on one library. Let me know what you think, meanwhile I'll look at what other libs have to offer. :)

hanupratap committed:
Changed lib to chartjs, removed unnecessary libs and files, formatted code

hanupratap commented:
Hey @rroohhh  :), I have tried chartjs library this time. This library is open-source and is performant. I have removed unnecessary dependencies and files. Also I have tested it for 5 min of data (approx. 1000 total points), works fine. Please let me know what you think. :)

hanupratap committed:
minor improvements

hanupratap committed:
minor performance improvement

hanupratap committed:
Fixed color in tooltip

hanupratap commented:
I have just added some minor improvements. Also, since we are using canvas element this time, we can download/view the current state of the graph by simply right-click -> save/view image.

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -0,0 +1,144 @@
+    return {
+      x_value: time_stamp,
+      y_value: [one_min, five_min, fifteen_min],
+      original_str: str,
```
unused

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -0,0 +1,144 @@
+
+  function extractLoadFromString(str = '') {
+    const list = str.split(',');
+    const time_stamp = list[0].trim();
```
This does not really work (atleast on my system). I get something like `xx:xx:xx up for x days x:xx` instead of just the `xx:xx:xx` part.

hanupratap replied:
got it

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -0,0 +1,144 @@
+
+  return (
+    <div>
+      <Button
```
The left edge of this button is cut-off for me.

rroohhh reviewed (changes requested):
chartjs seems be the best performing yet. However there are some points where it doesn't hold up that well
1. The zooming seems to act on a point granularity, meaning some amount of distance / rotation on the scroll wheel will make one point more / or less visible, regardless of the total number of visible points. This makes zooming basically impossible when there is a large number of points, as it gets very slow. I would expect zooming to work on a percentage base not a point base.
2. Panning is similarily strange, in that it works on a point granularity aswell. This means the panning speed is different depending on the number of points visible and does not match up with the cursor movement atall.
3. The x labels are handled in a very weird way. I would expect for a graph with time on the x axis to automatically choose suitable number of x ticks depending on the current zoom and label these with the corresponding time, completely independent of the data points.
4. This does not have a `x` value for the values atall. If some of the labels are the same, it renders the points on top of each other.
5. The popup when hovering over a point vanishes, when the chart is updated, making in essentially useless, as it is atmost visible for just 1 second.

rroohhh commented:
Oh also, having zoom on just the mousewheel makes it quite difficult to scroll the actual page.

hanupratap committed:
chartjs replaced with Dygraph charts

hanupratap committed:
Graph autoscale x-direction

hanupratap committed:
Added open source animation

hanupratap committed:
code format

hanupratap committed:
Added apertus animation

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -1,144 +1,80 @@
+    margin: '1rem',
+    marginBottom: '1.5rem',
+    display: 'flex',
+    fontFamily: 'Arial, Helvetica, sans-serif',
```
Is there a way to use the font from the theme instead of hardcoding it?

hanupratap replied:
Fixed

rroohhh commented on frontend/components/LoadGraph.jsx:
```diff
@@ -1,144 +1,80 @@
+  strokeWidth: 1.5,
+  title: chart_title,
+  titleHeight: 24,
+  xlabel: 'Time(s)',
```
sholud probably be only `Time`

hanupratap replied:
Fixed

rroohhh commented on frontend/util/animations/aperture.json:
```diff
@@ -0,0 +1 @@
+{"v":"4.6.9","fr":30,"ip":0,"op":300,"w":800,"h":600,"nm":"Aperture","ddd":0,"assets":[{"id":"comp_0","layers":[{"ddd":0,"ind":2,"ty":4,"nm":"Mask1","td":1,"ks":{"o":{"a":0,"k":100},"r":{"a":1,"k":[{"i":{"x":[0],"y":[1]},"o":{"x":[0.167],"y":[0.167]},"n":["0_1_0p167_0p167"],"t":0,"s":[0],"e":[-12]},{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":30 ...
```
Where is this coming from? If you have created it yourself, adding the source file aswell would be nice, otherwise, what license is it under?

hanupratap replied:
These are free Lottie-animations and comes under creative commons license. This is the [link ](https://lottiefiles.com/624-camera-aperture)to the animation, you can see the license at the end of the page. I have changed the colors to match the theme :)

rroohhh commented on frontend/util/animations/wifi_loading.json:
```diff
@@ -0,0 +1 @@
+{"v":"4.11.1","fr":29.9700012207031,"ip":0,"op":120.0000048877,"w":300,"h":300,"nm":"Text","ddd":0,"assets":[],"layers":[{"ddd":0,"ind":1,"ty":4,"nm":"internet Outlines","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":0,"k":0,"ix":10},"p":{"a":0,"k":[150,140,0],"ix":2},"a":{"a":0,"k":[50,42.5,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"ind":0,"ty":"sh"," ...
```
See above.

hanupratap replied:
This is the [link ](https://lottiefiles.com/8433-wifi-animation)to animation, I have changed the colors to match the theme :)

rroohhh reviewed (changes requested):
See inline.

For small screens, it would probably be nice to have the tooltip above the graph, as currently its just invisible below a certain width.

hanupratap committed:
changed tooltip position, removed hardcoded theme

hanupratap committed:
format code

hanupratap committed:
added accordion on sys-info, improved performanec

hanupratap committed:
format code

hanupratap committed:
small change in accordion

hanupratap commented:
Hey @rroohhh , I have added accordion on the sys-info page, also I have switched on the option of unmount when collapsed on the accordion which will improve performance.

hanupratap committed:
added dark mode

hanupratap committed:
color highlight fix
